### ISSUE TITLE
What's the intended workflow of the tarball cache?

### ISSUE THREAD

**Issue by wycats** [yarnpkg/yarn#393]
I understand how the feature works, but as I'm not using it, I'm not quite sure what the exact intended workflow is.

@bestander @skevy @kittens

**Comment by glenjamin** [comment]
I'm unclear from my limited use of yarn how much of a role the `resolved` field plays when installing from lockfile on a second machine - but the npm behaviour of always phoning home for this was very frustrating. It led me to always remove this field from the generated npm-shrinkwrap.

The specific flow we wanted was that I would have my development machine point at the public registry, but CI would go via a proxy.

I think the way @wycats describes storing a reference to the package source separately from the package location would help enable this workflow.

As a strawman, something along the lines of this might work:

The checked-in lock file states the expected source and a hash
```
abab@^1.0.0:
  version "1.0.3"
  source "registry"
  hash "b81de5f7274ec4e756d797cd834f303642724e5d"
```

Those sources would have default locations, and then separate environment-specific not-checked in config could override source locations - possibly giving an ordered list?

**Comment by mgcrea** [comment]
I've been testing custom yarn-cache folder for a few weeks, but I'm encountering a lot of `Tarball is not in network and can not be located in cache` errors (on gitlab-ci, etc.) that would only be solved with a `yarn --no-lockfile` install. By any chance, would you have encountered such errors?

```
"error \"bunyan-1.8.4.tgz\": Tarball is not in network and can not be located in cache (\"/srv/player/.yarn-cache/bunyan-1.8.4.tgz\")", "stdout": "yarn install v0.17.0\n[1/4] Resolving packages...\n[2/4] Fetching packages...\ninfo Visit https://yarnpkg.com/en/docs/cli/install for documentation about this command."
```

**Comment by UnrememberMe** [comment]
Somewhat related but maybe a little stray from the topic, is there any thoughts about dealing with node module installation scripts? There are plenty of node modules download additional codes during installation and thus the results of `yarn install` is still non-repeatable even with the offline mirror.

**Comment by bestander** [comment]
The problem is that Node.js install scripts can execute any bash script, there is no way to reliably achieve offline mode without authors' cooperation.
The only thing that we can do is not use such modules or help the authors to think of offline mode.

**Comment by UnrememberMe** [comment]
How about caching the post-install results instead of pre-install results?  I understand that will not work well with any codes with native platform dependencies, but that is not a problem any of our current solutions address either.

The worst can happen is that we can not use those badly behaving modules, which will be as bad as current situation.

**Comment by bestander** [comment]
Then it is as good as saving node_modules somewhere, for example, checking them into source control.

**Comment by UnrememberMe** [comment]
I agree that will be equivalent, with added benefits that offline mirror currently provides:
- only store one checksumed tgz file vs. hundreds (if not thousands) of files
- the stored node_modules can be shared amount different projects;
- code review for updating a dependency remain reasonably sane;

**Comment by bestander** [comment]
@UnrememberMe, better discuss this in a separate issue/RFC

**Comment by UnrememberMe** [comment]
Agreed.  Will open a separate issue/RFC.

**Comment by gregsheremeta** [comment]
@UnrememberMe did you happen to open one? Could you link to it?

**Comment by UnrememberMe** [comment]
I have started but have not finished the RFC yet.  I should submit the pull request for RFC no later than Thursday 2/16/17. @gregsheremeta

**Comment by jackhamburger** [comment]
Is there a concern about the offline mirror will start to bloat if it is stored in source control? Every minor version change will leave the old .tgz files. 
Is there a plan for a clean up command that empties the offline mirror of module .tgzs that are not used in the yarn.lock?

**Comment by bestander** [comment]
Yeah, there is an RFC for that already.
An opt-in cleanup feature is coming

**Comment by UnrememberMe** [comment]
@gregsheremeta The RFC was posted on Feb 16, 2017.

**Comment by gregsheremeta** [comment]
@UnrememberMe do you have a link to it? I don't see it in https://github.com/yarnpkg/rfcs

**Comment by UnrememberMe** [comment]
@gregsheremeta I updated the title  for https://github.com/yarnpkg/rfcs/pull/50  The initial RFC title was not correct.

**Comment by Vanuan** [comment]
That's something I'm interested in too. I'll leave my thoughts here:

AFAIK, current yarn workflow is the following:

1. Set `$HOME` to isolated location for each build
2. Use `yarn install --pure-lock-file` to make sure people check in `yarn.lock`
3. Use `yarn add/remove` commands to change package.json and yarn.lock
4. When there are conflicts, remove `yarn.lock` and do clean yarn install :( <--- not sure if it's correct

Each step could be improved:

1. Yarn should be able to use some kind of cache corruption prevention by default (is it possible to do using a file in `$HOME`?)
2. Pure lock file should be used by default, otherwise people wouldn't commit `yarn.lock` and conflicts are inevitable
3. Some people edit package.json manually, but it isn't easy to edit yarn.lock, so yarn install --pure-lock-file should still install packages present in package.json, but missing from yarn.lock
4. Some kind of smart `yarn git-resolve` would be nice

How is it related to tarball cache feature? Not much. Intended use case of it is to completely avoid yarn install during CI and saving all dependencies in the repository.

As we can see, local tarball cache would resolve all CI issues but somewhat complicate developer workflow issues.

Ideally, we'd want online repository but cacheable.
Yarn already has `.cache` thus tarball cache doesn't look useful.
I.e. tarball cache wouldn't be needed if each `yarn install` updated $HOME/ cache without corruption. And committing dependencies to git wouldn't be needed if it was guaranteed that npm registry is always online in case cache is corrupted.

TL;DR It would be nice if workflow wouldn't change while achieving performance and stability improvement.

**Comment by raix** [comment]
@Vanuan in case of merge conflicts theres a merged pr making `yarn install` solve merge conflicts (https://github.com/yarnpkg/yarn/pull/3544)

**Comment by Vanuan** [comment]
Yeap, I'm aware of that. That comment was back in April.

**Comment by BYK** [comment]
Closing this issue since it seems to be resolved, mostly by #2970 but possibly with other PRs.

Please create a new issue if you want to propose more features of fixes around this.

**Comment by varemenos** [comment]
The past part of [this](https://yarnpkg.com/blog/2016/11/24/offline-mirror/) blog post is referencing this ticket. Is it still valid or should the blog post be updated?
